src/pages/sicp/subcomponents/chatbot/ChatBox.tsx

```diff
@@ -0,0 +1,183 @@
+  const chatRef = React.useRef<HTMLDivElement | null>(null);
+  const [isLoading, setIsLoading] = React.useState(false);
+  const [messages, setMessages] = React.useState<{ role: 'user' | 'bot'; content: string[] }[]>([
+    { content: ['Ask me something about this paragraph!'], role: 'bot' }
+  ]);
+  const [userInput, setUserInput] = React.useState<string>('');
+  const [contentHistory, setContentHistory] = React.useState<Array<string>>([]);
+  const [roleHistory, setRoleHistory] = React.useState<Array<string>>([]);
+
+  const handleUserInput = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setUserInput(event.target.value);
+  };
+
+  // To get code snippets
+  const codeBlocks = (temp: string) => {
+    return temp.split('```');
+  };
+
+  const text = () => {
+    return '\n(2) Here is the paragraph:\n' + getText();
+  };
+
+  const section = () => {
+    const sectionNumber = getSection();
+    return parseInt(sectionNumber.charAt(0), 10) > 3
+      ? '\n(1) There is no section summary for this section. Please answer the question based on the following paragraph\n'
+      : '\n(1) Here is the summary of this section:\n' + SICPNotes[getSection()];
+  };
+
+  function getPrompt() {
+    const prompt =
+      'You are a competent tutor, assisting a student who is learning computer science following the textbook "Structure and Interpretation of Computer Programs,' +
+      'JavaScript edition". The student request is about a paragraph of the book. The request may be a follow-up request to a request that was posed to you' +
+      'previously.\n' +
+      'What follows are:\n' +
+      '(1) the summary of section (2) the full paragraph. Please answer the student request,' +
+      'not the requests of the history. If the student request is not related to the book, ask them to ask questions that are related to the book. Donot say that I provide you text\n\n' +
+      section() +
+      text();
+    return prompt;
+  }
+
+  const { accessToken, refreshToken } = useSession();
+
+  const sendMessage = async () => {
+    if (userInput.trim() !== '') {
+      // clean the input immediately after the user sends the message so that the user would not feel the lag
+      const _userInput = userInput;
+      const _messages = messages;
+      setUserInput('');
+      const blocks = codeBlocks(_userInput);
+      setMessages([...messages, { role: 'user', content: blocks }]);
+      setIsLoading(true);
+
+      const prompt = getPrompt();
+      const payload = [{ role: 'system', content: prompt }];
+      for (let i = 0; i < contentHistory.length; i++) {
+        payload.push({ role: roleHistory[i], content: contentHistory[i] });
+      }
+      payload.push({ role: 'user', content: _userInput });
+      const tokens = { accessToken, refreshToken };
+      const response = await request(`chat`, 'POST', {
+        ...tokens,
+        body: { json: payload }
+      });
+      const text = await response?.text();
```

Comment:
Is it possible to move this to our Saga instead?

Reply:
Will work on that tonight.

Reply:
Could you provide some guidance on how this can be done?

Given how strongly coupled this is request is to the other local state, the furthest I can think of doing is like this: 
```
export async function chatComplete(tokens: Tokens, payload:[{role:string, content:string}]): Promise<string | undefined> {
      const response = await request(`chat`, 'POST', {
        ...tokens,
        body: { json: payload }
      });
      const text = await response?.text();
      return text;
}
```
But then I am confused by the backend saga(I need to pass in tokes). What are "yield" , "function*" and "ReturnType<typeof actions.xxx>" ? Do I need to transfer to that?

Reply:
@RichDom2185 Hi, I have made the requested change in the commit "[Migrate request to Saga](https://github.com/source-academy/frontend/pull/2813/commits/6496254851bf2379b57f7dc3a67d4ee7a1dddbc2)". Is this what intended for? Thanks.